File: src/compiler.c
/*
 * compiler.c — Compiles AST to bytecode for the stack VM.
 *
 * Key design for recursion:
 *   - Each function is compiled to bytecode with OP_ENTER (allocate locals).
 *   - OP_CALL saves the frame, sets a new frame pointer.
 *   - Variables are accessed via OP_LOAD/OP_STORE relative to frame pointer.
 *   - Recursive calls naturally get new frames with isolated locals.
 */

#include "compiler.h"

/* ── Compiler state ─────────────────────────────────────── */
typedef struct {
    char name[MAX_IDENT_LEN];
    int  slot;      /* index relative to frame pointer */
    int  depth;     /* scope depth                     */
} Local;

typedef struct {
    char name[MAX_IDENT_LEN];
    int  address;   /* bytecode address (-1 = not yet compiled) */
    int  n_params;
} FuncInfo;

typedef struct {
    int position;   /* index in code[] to patch */
    int func_idx;   /* index into functions[]   */
} Patch;

typedef struct {
    Program  *prog;

    /* Current function context */
    Local     locals[MAX_LOCALS];
    int       local_count;
    int       scope_depth;
    int       n_params;       /* params in current function */
    int       n_locals;       /* locals (non-param) count   */

    /* Function table */
    FuncInfo  functions[MAX_FUNCTIONS];
    int       func_count;

    /* Forward-reference patches */
    Patch     patches[MAX_PATCHES];
    int       patch_count;
} Compiler;

/* ── Emit helpers ───────────────────────────────────────── */
static void emit(Compiler *c, int val) {
    if (c->prog->count >= MAX_CODE)
        error_exit(0, "bytecode overflow (program too large)");
    c->prog->code[c->prog->count++] = val;
}

static int emit_placeholder(Compiler *c) {
    int pos = c->prog->count;
    emit(c, 0);
    return pos;
}

static void patch_at(Compiler *c, int pos, int val) {
    c->prog->code[pos] = val;
}

static int here(Compiler *c) { return c->prog->count; }

/* ── Scope management ───────────────────────────────────── */
static void begin_scope(Compiler *c) { c->scope_depth++; }

static void end_scope(Compiler *c) {
    while (c->local_count > 0 &&
           c->locals[c->local_count - 1].depth == c->scope_depth) {
        c->local_count--;
    }
    c->scope_depth--;
}

static int add_local(Compiler *c, const char *name, int line) {
    /* Check for duplicate in same scope */
    for (int i = c->local_count - 1; i >= 0; i--) {
        if (c->locals[i].depth < c->scope_depth) break;
        if (strcmp(c->locals[i].name, name) == 0)
            error_exit(line, "variable '%s' already declared in this scope", name);
    }
    if (c->local_count >= MAX_LOCALS)
        error_exit(line, "too many local variables");

    int slot = c->n_params + c->n_locals;
    Local *loc = &c->locals[c->local_count++];
    strncpy(loc->name, name, MAX_IDENT_LEN - 1);
    loc->name[MAX_IDENT_LEN - 1] = '\0';
    loc->slot  = slot;
    loc->depth = c->scope_depth;
    c->n_locals++;
    return slot;
}

static int resolve_local(Compiler *c, const char *name) {
    for (int i = c->local_count - 1; i >= 0; i--) {
        if (strcmp(c->locals[i].name, name) == 0)
            return c->locals[i].slot;
    }
    return -1;
}

/* ── Function table ─────────────────────────────────────── */
static int find_func(Compiler *c, const char *name) {
    for (int i = 0; i < c->func_count; i++)
        if (strcmp(c->functions[i].name, name) == 0) return i;
    return -1;
}

static int register_func(Compiler *c, const char *name, int n_params, int line) {
    if (c->func_count >= MAX_FUNCTIONS)
        error_exit(line, "too many functions");
    int idx = c->func_count++;
    strncpy(c->functions[idx].name, name, MAX_IDENT_LEN - 1);
    c->functions[idx].address  = -1;
    c->functions[idx].n_params = n_params;
    return idx;
}

/* ── Compile expressions ────────────────────────────────── */
static void compile_expr(Compiler *c, ASTNode *n);
static void compile_stmt(Compiler *c, ASTNode *n);

static void compile_expr(Compiler *c, ASTNode *n) {
    switch (n->type) {
    case NODE_NUMBER:
        emit(c, OP_CONST);
        emit(c, n->int_value);
        break;

    case NODE_IDENT: {
        int slot = resolve_local(c, n->name);
        if (slot < 0)
            error_exit(n->line, "undefined variable '%s'", n->name);
        emit(c, OP_LOAD);
        emit(c, slot);
        break;
    }

    case NODE_ASSIGN: {
        int slot = resolve_local(c, n->name);
        if (slot < 0)
            error_exit(n->line, "undefined variable '%s'", n->name);
        compile_expr(c, n->expr);
        emit(c, OP_STORE);
        emit(c, slot);
        /* Assignment is an expression: leave the value on the stack */
        emit(c, OP_LOAD);
        emit(c, slot);
        break;
    }

    case NODE_BINARY:
        compile_expr(c, n->left);
        compile_expr(c, n->right);
        switch (n->op) {
            case TOKEN_PLUS:    emit(c, OP_ADD); break;
            case TOKEN_MINUS:   emit(c, OP_SUB); break;
            case TOKEN_STAR:    emit(c, OP_MUL); break;
            case TOKEN_SLASH:   emit(c, OP_DIV); break;
            case TOKEN_PERCENT: emit(c, OP_MOD); break;
            case TOKEN_EQ:      emit(c, OP_EQ);  break;
            case TOKEN_NEQ:     emit(c, OP_NEQ); break;
            case TOKEN_LT:      emit(c, OP_LT);  break;
            case TOKEN_GT:      emit(c, OP_GT);  break;
            case TOKEN_LTE:     emit(c, OP_LTE); break;
            case TOKEN_GTE:     emit(c, OP_GTE); break;
            case TOKEN_AND:     emit(c, OP_AND); break;
            case TOKEN_OR:      emit(c, OP_OR);  break;
            default:
                error_exit(n->line, "unknown binary operator");
        }
        break;

    case NODE_UNARY:
        compile_expr(c, n->left);
        switch (n->op) {
            case TOKEN_MINUS: emit(c, OP_NEG); break;
            case TOKEN_NOT:   emit(c, OP_NOT); break;
            default:
                error_exit(n->line, "unknown unary operator");
        }
        break;

    case NODE_CALL: {
        /* Check for built-in functions */
        if (strcmp(n->name, "print") == 0) {
            if (n->item_count != 1)
                error_exit(n->line, "print() expects exactly 1 argument");
            compile_expr(c, n->items[0]);
            emit(c, OP_PRINT);
            break;
        }
        if (strcmp(n->name, "scan") == 0) {
            if (n->item_count != 0)
                error_exit(n->line, "scan() expects no arguments");
            emit(c, OP_SCAN);
            break;
        }

        /* User-defined function call */
        int fi = find_func(c, n->name);
        if (fi < 0)
            error_exit(n->line, "undefined function '%s'", n->name);
        if (n->item_count != c->functions[fi].n_params)
            error_exit(n->line, "'%s' expects %d args, got %d",
                       n->name, c->functions[fi].n_params, n->item_count);

        /* Push arguments */
        for (int i = 0; i < n->item_count; i++)
            compile_expr(c, n->items[i]);

        /* Emit CALL: address will be patched if forward reference */
        emit(c, OP_CALL);
        if (c->functions[fi].address >= 0) {
            emit(c, c->functions[fi].address);
        } else {
            /* Forward reference — add to patch list */
            if (c->patch_count >= MAX_PATCHES)
                error_exit(n->line, "too many forward references");
            c->patches[c->patch_count].position = c->prog->count;
            c->patches[c->patch_count].func_idx = fi;
            c->patch_count++;
            emit_placeholder(c);
        }
        emit(c, n->item_count);
        break;
    }

    default:
        error_exit(n->line, "invalid expression node type %d", n->type);
    }
}

/* ── Compile statements ─────────────────────────────────── */
static void compile_stmt(Compiler *c, ASTNode *n) {
    switch (n->type) {
    case NODE_EXPR_STMT:
        compile_expr(c, n->expr);
        emit(c, OP_POP);   /* discard expression result */
        break;

    case NODE_VAR_DECL: {
        int slot = add_local(c, n->name, n->line);
        if (n->init) {
            compile_expr(c, n->init);
        } else {
            emit(c, OP_CONST);
            emit(c, 0);    /* default initialise to 0 */
        }
        emit(c, OP_STORE);
        emit(c, slot);
        break;
    }

    case NODE_BLOCK:
        begin_scope(c);
        for (int i = 0; i < n->item_count; i++)
            compile_stmt(c, n->items[i]);
        end_scope(c);
        break;

    case NODE_IF: {
        compile_expr(c, n->cond);
        emit(c, OP_JZ);
        int else_jump = emit_placeholder(c);

        compile_stmt(c, n->body);

        if (n->else_body) {
            emit(c, OP_JMP);
            int end_jump = emit_placeholder(c);
            patch_at(c, else_jump, here(c));
            compile_stmt(c, n->else_body);
            patch_at(c, end_jump, here(c));
        } else {
            patch_at(c, else_jump, here(c));
        }
        break;
    }

    case NODE_WHILE: {
        int loop_start = here(c);
        compile_expr(c, n->cond);
        emit(c, OP_JZ);
        int exit_jump = emit_placeholder(c);

        compile_stmt(c, n->body);
        emit(c, OP_JMP);
        emit(c, loop_start);

        patch_at(c, exit_jump, here(c));
        break;
    }

    case NODE_FOR: {
        begin_scope(c);

        /* Init */
        if (n->init) {
            if (n->init->type == NODE_VAR_DECL)
                compile_stmt(c, n->init);
            else {
                compile_expr(c, n->init);
                emit(c, OP_POP);
            }
        }

        int loop_start = here(c);

        /* Condition */
        if (n->cond) {
            compile_expr(c, n->cond);
        } else {
            emit(c, OP_CONST);
            emit(c, 1); /* infinite loop if no condition */
        }
        emit(c, OP_JZ);
        int exit_jump = emit_placeholder(c);

        /* Body */
        compile_stmt(c, n->body);

        /* Update */
        if (n->update) {
            compile_expr(c, n->update);
            emit(c, OP_POP);
        }

        emit(c, OP_JMP);
        emit(c, loop_start);

        patch_at(c, exit_jump, here(c));
        end_scope(c);
        break;
    }

    case NODE_RETURN:
        if (n->expr)
            compile_expr(c, n->expr);
        else {
            emit(c, OP_CONST);
            emit(c, 0);
        }
        emit(c, OP_RET);
        break;

    default:
        error_exit(n->line, "unexpected statement node type %d", n->type);
    }
}

/* ── Compile a function definition ──────────────────────── */
static void compile_func(Compiler *c, ASTNode *fn) {
    int fi = find_func(c, fn->name);

    /* Set function address to current position */
    c->functions[fi].address = here(c);

    /* Reset local context */
    c->local_count = 0;
    c->scope_depth = 0;
    c->n_params    = fn->item_count;
    c->n_locals    = 0;

    /* Register parameters as locals (slot 0..n-1) */
    begin_scope(c);
    for (int i = 0; i < fn->item_count; i++) {
        ASTNode *p = fn->items[i];
        if (c->local_count >= MAX_LOCALS)
            error_exit(p->line, "too many parameters");
        Local *loc = &c->locals[c->local_count++];
        strncpy(loc->name, p->name, MAX_IDENT_LEN - 1);
        loc->name[MAX_IDENT_LEN - 1] = '\0';
        loc->slot  = i;
        loc->depth = c->scope_depth;
    }

    /* Emit ENTER with placeholder for local count */
    emit(c, OP_ENTER);
    int locals_pos = emit_placeholder(c);

    /* Compile body */
    if (fn->body->type == NODE_BLOCK) {
        /* Compile block contents directly (scope already open) */
        for (int i = 0; i < fn->body->item_count; i++)
            compile_stmt(c, fn->body->items[i]);
    } else {
        compile_stmt(c, fn->body);
    }

    /* Safety net: if function doesn't explicitly return, push 0 and return */
    emit(c, OP_CONST);
    emit(c, 0);
    emit(c, OP_RET);

    /* Patch ENTER with actual local count */
    patch_at(c, locals_pos, c->n_locals);

    end_scope(c);
}

/* ── Main compile entry point ───────────────────────────── */
int compiler_compile(ASTNode *ast, Program *prog) {
    Compiler c;
    memset(&c, 0, sizeof(c));
    c.prog = prog;
    prog->count = 0;
    prog->entry = -1;

    /* Emit preamble: CALL main, HALT */
    emit(&c, OP_CALL);
    int main_addr_pos = emit_placeholder(&c);
    emit(&c, 0);   /* 0 args */
    emit(&c, OP_POP); /* discard main's return value */
    emit(&c, OP_HALT);

    /* Pass 1: register all function signatures */
    for (int i = 0; i < ast->item_count; i++) {
        ASTNode *fn = ast->items[i];
        if (fn->type != NODE_FUNC_DEF)
            error_exit(fn->line, "only function definitions allowed at top level");
        if (find_func(&c, fn->name) >= 0)
            error_exit(fn->line, "function '%s' already defined", fn->name);
        register_func(&c, fn->name, fn->item_count, fn->line);
    }

    /* Pass 2: compile all function bodies */
    for (int i = 0; i < ast->item_count; i++)
        compile_func(&c, ast->items[i]);

    /* Resolve forward references */
    for (int i = 0; i < c.patch_count; i++) {
        int fi   = c.patches[i].func_idx;
        int pos  = c.patches[i].position;
        if (c.functions[fi].address < 0)
            error_exit(0, "function '%s' declared but never defined",
                       c.functions[fi].name);
        patch_at(&c, pos, c.functions[fi].address);
    }

    /* Patch main() address in preamble */
    int main_idx = find_func(&c, "main");
    if (main_idx < 0)
        error_exit(0, "no 'main' function defined");
    patch_at(&c, main_addr_pos, c.functions[main_idx].address);
    prog->entry = 0;

    return 0;
}

/* ── Disassembler (for debugging) ───────────────────────── */
static const char *op_name(int op) {
    switch (op) {
        case OP_CONST: return "CONST";   case OP_POP:   return "POP";
        case OP_ADD:   return "ADD";     case OP_SUB:   return "SUB";
        case OP_MUL:   return "MUL";     case OP_DIV:   return "DIV";
        case OP_MOD:   return "MOD";     case OP_NEG:   return "NEG";
        case OP_EQ:    return "EQ";      case OP_NEQ:   return "NEQ";
        case OP_LT:    return "LT";      case OP_GT:    return "GT";
        case OP_LTE:   return "LTE";     case OP_GTE:   return "GTE";
        case OP_AND:   return "AND";     case OP_OR:    return "OR";
        case OP_NOT:   return "NOT";
        case OP_LOAD:  return "LOAD";    case OP_STORE: return "STORE";
        case OP_JMP:   return "JMP";     case OP_JZ:    return "JZ";
        case OP_CALL:  return "CALL";    case OP_RET:   return "RET";
        case OP_ENTER: return "ENTER";
        case OP_PRINT: return "PRINT";   case OP_SCAN:  return "SCAN";
        case OP_HALT:  return "HALT";
        default:       return "???";
    }
}

void compiler_disassemble(const Program *prog) {
    printf("=== Bytecode (%d instructions) ===\n", prog->count);
    int i = 0;
    while (i < prog->count) {
        int op = prog->code[i];
        printf("%04d  %-8s", i, op_name(op));
        switch (op) {
            case OP_CONST: case OP_LOAD: case OP_STORE:
            case OP_JMP:   case OP_JZ:   case OP_ENTER:
                printf(" %d", prog->code[i + 1]);
                i += 2;
                break;
            case OP_CALL:
                printf(" addr=%d nargs=%d", prog->code[i+1], prog->code[i+2]);
                i += 3;
                break;
            default:
                i += 1;
                break;
        }
        printf("\n");
    }
    printf("=================================\n");
}

File: desktop/compiler/compiler.js
/**
 * compiler.js — Direct port of src/compiler.c
 * Compiles AST to bytecode for the stack VM.
 */

const { TokenType, CompileError, tokenTypeName } = require('./lexer');
const { NodeType } = require('./ast');

const OpCode = Object.freeze({
    OP_CONST:  0,
    OP_POP:    1,
    OP_ADD:    2,
    OP_SUB:    3,
    OP_MUL:    4,
    OP_DIV:    5,
    OP_MOD:    6,
    OP_NEG:    7,
    OP_EQ:     8,
    OP_NEQ:    9,
    OP_LT:    10,
    OP_GT:    11,
    OP_LTE:   12,
    OP_GTE:   13,
    OP_AND:   14,
    OP_OR:    15,
    OP_NOT:   16,
    OP_LOAD:  17,
    OP_STORE: 18,
    OP_JMP:   19,
    OP_JZ:    20,
    OP_CALL:  21,
    OP_RET:   22,
    OP_ENTER: 23,
    OP_PRINT: 24,
    OP_SCAN:  25,
    OP_HALT:  26,
});

const OP_NAMES = {
    [OpCode.OP_CONST]:  'CONST',
    [OpCode.OP_POP]:    'POP',
    [OpCode.OP_ADD]:    'ADD',
    [OpCode.OP_SUB]:    'SUB',
    [OpCode.OP_MUL]:    'MUL',
    [OpCode.OP_DIV]:    'DIV',
    [OpCode.OP_MOD]:    'MOD',
    [OpCode.OP_NEG]:    'NEG',
    [OpCode.OP_EQ]:     'EQ',
    [OpCode.OP_NEQ]:    'NEQ',
    [OpCode.OP_LT]:     'LT',
    [OpCode.OP_GT]:     'GT',
    [OpCode.OP_LTE]:    'LTE',
    [OpCode.OP_GTE]:    'GTE',
    [OpCode.OP_AND]:    'AND',
    [OpCode.OP_OR]:     'OR',
    [OpCode.OP_NOT]:    'NOT',
    [OpCode.OP_LOAD]:   'LOAD',
    [OpCode.OP_STORE]:  'STORE',
    [OpCode.OP_JMP]:    'JMP',
    [OpCode.OP_JZ]:     'JZ',
    [OpCode.OP_CALL]:   'CALL',
    [OpCode.OP_RET]:    'RET',
    [OpCode.OP_ENTER]:  'ENTER',
    [OpCode.OP_PRINT]:  'PRINT',
    [OpCode.OP_SCAN]:   'SCAN',
    [OpCode.OP_HALT]:   'HALT',
};

function opName(op) {
    return OP_NAMES[op] || '???';
}

class Program {
    constructor() {
        this.code = [];
        this.count = 0;
        this.entry = -1;
    }
}

const MAX_LOCALS = 256;
const MAX_FUNCTIONS = 64;
const MAX_PATCHES = 512;

class Compiler {
    constructor() {
        this.prog = new Program();
        this.locals = [];
        this.localCount = 0;
        this.scopeDepth = 0;
        this.nParams = 0;
        this.nLocals = 0;
        this.functions = [];
        this.funcCount = 0;
        this.patches = [];
        this.patchCount = 0;

        // For semantic info visualization
        this.semanticInfo = {
            functions: [],
            scopes: [],
        };
    }

    // ── Emit helpers ──────────────────────────────────
    emit(val) {
        this.prog.code.push(val);
        this.prog.count++;
    }

    emitPlaceholder() {
        const pos = this.prog.count;
        this.emit(0);
        return pos;
    }

    patchAt(pos, val) {
        this.prog.code[pos] = val;
    }

    here() { return this.prog.count; }

    // ── Scope management ──────────────────────────────
    beginScope() { this.scopeDepth++; }

    endScope() {
        while (this.localCount > 0 &&
               this.locals[this.localCount - 1].depth === this.scopeDepth) {
            this.localCount--;
        }
        this.scopeDepth--;
    }

    addLocal(name, line) {
        for (let i = this.localCount - 1; i >= 0; i--) {
            if (this.locals[i].depth < this.scopeDepth) break;
            if (this.locals[i].name === name) {
                throw new CompileError(line, `variable '${name}' already declared in this scope`);
            }
        }
        if (this.localCount >= MAX_LOCALS) {
            throw new CompileError(line, 'too many local variables');
        }

        const slot = this.nParams + this.nLocals;
        this.locals[this.localCount++] = { name, slot, depth: this.scopeDepth };
        this.nLocals++;
        return slot;
    }

    resolveLocal(name) {
        for (let i = this.localCount - 1; i >= 0; i--) {
            if (this.locals[i].name === name) return this.locals[i].slot;
        }
        return -1;
    }

    // ── Function table ────────────────────────────────
    findFunc(name) {
        for (let i = 0; i < this.funcCount; i++) {
            if (this.functions[i].name === name) return i;
        }
        return -1;
    }

    registerFunc(name, nParams, line) {
        if (this.funcCount >= MAX_FUNCTIONS) {
            throw new CompileError(line, 'too many functions');
        }
        const idx = this.funcCount++;
        this.functions[idx] = { name, address: -1, nParams };
        return idx;
    }

    // ── Compile expressions ──────────────────────────
    compileExpr(n) {
        switch (n.type) {
            case NodeType.NODE_NUMBER:
                this.emit(OpCode.OP_CONST);
                this.emit(n.int_value);
                break;

            case NodeType.NODE_IDENT: {
                const slot = this.resolveLocal(n.name);
                if (slot < 0) throw new CompileError(n.line, `undefined variable '${n.name}'`);
                this.emit(OpCode.OP_LOAD);
                this.emit(slot);
                break;
            }

            case NodeType.NODE_ASSIGN: {
                const slot = this.resolveLocal(n.name);
                if (slot < 0) throw new CompileError(n.line, `undefined variable '${n.name}'`);
                this.compileExpr(n.expr);
                this.emit(OpCode.OP_STORE);
                this.emit(slot);
                this.emit(OpCode.OP_LOAD);
                this.emit(slot);
                break;
            }

            case NodeType.NODE_BINARY:
                this.compileExpr(n.left);
                this.compileExpr(n.right);
                switch (n.op) {
                    case TokenType.TOKEN_PLUS:    this.emit(OpCode.OP_ADD); break;
                    case TokenType.TOKEN_MINUS:   this.emit(OpCode.OP_SUB); break;
                    case TokenType.TOKEN_STAR:    this.emit(OpCode.OP_MUL); break;
                    case TokenType.TOKEN_SLASH:   this.emit(OpCode.OP_DIV); break;
                    case TokenType.TOKEN_PERCENT: this.emit(OpCode.OP_MOD); break;
                    case TokenType.TOKEN_EQ:      this.emit(OpCode.OP_EQ);  break;
                    case TokenType.TOKEN_NEQ:     this.emit(OpCode.OP_NEQ); break;
                    case TokenType.TOKEN_LT:      this.emit(OpCode.OP_LT);  break;
                    case TokenType.TOKEN_GT:      this.emit(OpCode.OP_GT);  break;
                    case TokenType.TOKEN_LTE:     this.emit(OpCode.OP_LTE); break;
                    case TokenType.TOKEN_GTE:     this.emit(OpCode.OP_GTE); break;
                    case TokenType.TOKEN_AND:     this.emit(OpCode.OP_AND); break;
                    case TokenType.TOKEN_OR:      this.emit(OpCode.OP_OR);  break;
                    default:
                        throw new CompileError(n.line, 'unknown binary operator');
                }
                break;

            case NodeType.NODE_UNARY:
                this.compileExpr(n.left);
                switch (n.op) {
                    case TokenType.TOKEN_MINUS: this.emit(OpCode.OP_NEG); break;
                    case TokenType.TOKEN_NOT:   this.emit(OpCode.OP_NOT); break;
                    default:
                        throw new CompileError(n.line, 'unknown unary operator');
                }
                break;

            case NodeType.NODE_CALL: {
                if (n.name === 'print') {
                    if (n.items.length !== 1)
                        throw new CompileError(n.line, 'print() expects exactly 1 argument');
                    this.compileExpr(n.items[0]);
                    this.emit(OpCode.OP_PRINT);
                    break;
                }
                if (n.name === 'scan') {
                    if (n.items.length !== 0)
                        throw new CompileError(n.line, 'scan() expects no arguments');
                    this.emit(OpCode.OP_SCAN);
                    break;
                }

                const fi = this.findFunc(n.name);
                if (fi < 0) throw new CompileError(n.line, `undefined function '${n.name}'`);
                if (n.items.length !== this.functions[fi].nParams)
                    throw new CompileError(n.line,
                        `'${n.name}' expects ${this.functions[fi].nParams} args, got ${n.items.length}`);

                for (let i = 0; i < n.items.length; i++) {
                    this.compileExpr(n.items[i]);
                }

                this.emit(OpCode.OP_CALL);
                if (this.functions[fi].address >= 0) {
                    this.emit(this.functions[fi].address);
                } else {
                    if (this.patchCount >= MAX_PATCHES)
                        throw new CompileError(n.line, 'too many forward references');
                    this.patches[this.patchCount++] = {
                        position: this.prog.count,
                        funcIdx: fi,
                    };
                    this.emitPlaceholder();
                }
                this.emit(n.items.length);
                break;
            }

            default:
                throw new CompileError(n.line, `invalid expression node type ${n.type}`);
        }
    }

    // ── Compile statements ────────────────────────────
    compileStmt(n) {
        switch (n.type) {
            case NodeType.NODE_EXPR_STMT:
                this.compileExpr(n.expr);
                this.emit(OpCode.OP_POP);
                break;

            case NodeType.NODE_VAR_DECL: {
                const slot = this.addLocal(n.name, n.line);
                if (n.init) {
                    this.compileExpr(n.init);
                } else {
                    this.emit(OpCode.OP_CONST);
                    this.emit(0);
                }
                this.emit(OpCode.OP_STORE);
                this.emit(slot);
                break;
            }

            case NodeType.NODE_BLOCK:
                this.beginScope();
                for (let i = 0; i < n.items.length; i++) {
                    this.compileStmt(n.items[i]);
                }
                this.endScope();
                break;

            case NodeType.NODE_IF: {
                this.compileExpr(n.cond);
                this.emit(OpCode.OP_JZ);
                const elseJump = this.emitPlaceholder();

                this.compileStmt(n.body);

                if (n.else_body) {
                    this.emit(OpCode.OP_JMP);
                    const endJump = this.emitPlaceholder();
                    this.patchAt(elseJump, this.here());
                    this.compileStmt(n.else_body);
                    this.patchAt(endJump, this.here());
                } else {
                    this.patchAt(elseJump, this.here());
                }
                break;
            }

            case NodeType.NODE_WHILE: {
                const loopStart = this.here();
                this.compileExpr(n.cond);
                this.emit(OpCode.OP_JZ);
                const exitJump = this.emitPlaceholder();

                this.compileStmt(n.body);
                this.emit(OpCode.OP_JMP);
                this.emit(loopStart);

                this.patchAt(exitJump, this.here());
                break;
            }

            case NodeType.NODE_FOR: {
                this.beginScope();

                if (n.init) {
                    if (n.init.type === NodeType.NODE_VAR_DECL) {
                        this.compileStmt(n.init);
                    } else {
                        this.compileExpr(n.init);
                        this.emit(OpCode.OP_POP);
                    }
                }

                const loopStart = this.here();

                if (n.cond) {
                    this.compileExpr(n.cond);
                } else {
                    this.emit(OpCode.OP_CONST);
                    this.emit(1);
                }
                this.emit(OpCode.OP_JZ);
                const exitJump = this.emitPlaceholder();

                this.compileStmt(n.body);

                if (n.update) {
                    this.compileExpr(n.update);
                    this.emit(OpCode.OP_POP);
                }

                this.emit(OpCode.OP_JMP);
                this.emit(loopStart);

                this.patchAt(exitJump, this.here());
                this.endScope();
                break;
            }

            case NodeType.NODE_RETURN:
                if (n.expr) {
                    this.compileExpr(n.expr);
                } else {
                    this.emit(OpCode.OP_CONST);
                    this.emit(0);
                }
                this.emit(OpCode.OP_RET);
                break;

            default:
                throw new CompileError(n.line, `unexpected statement node type ${n.type}`);
        }
    }

    // ── Compile function ──────────────────────────────
    compileFunc(fn) {
        const fi = this.findFunc(fn.name);
        this.functions[fi].address = this.here();

        this.localCount = 0;
        this.scopeDepth = 0;
        this.nParams = fn.items.length;
        this.nLocals = 0;

        this.beginScope();
        for (let i = 0; i < fn.items.length; i++) {
            const p = fn.items[i];
            this.locals[this.localCount++] = {
                name: p.name,
                slot: i,
                depth: this.scopeDepth,
            };
        }

        this.emit(OpCode.OP_ENTER);
        const localsPos = this.emitPlaceholder();

        if (fn.body.type === NodeType.NODE_BLOCK) {
            for (let i = 0; i < fn.body.items.length; i++) {
                this.compileStmt(fn.body.items[i]);
            }
        } else {
            this.compileStmt(fn.body);
        }

        this.emit(OpCode.OP_CONST);
        this.emit(0);
        this.emit(OpCode.OP_RET);

        this.patchAt(localsPos, this.nLocals);

        // Record semantic info
        this.semanticInfo.functions.push({
            name: fn.name,
            address: this.functions[fi].address,
            params: fn.items.map(p => p.name),
            localCount: this.nLocals,
            returnType: tokenTypeName(fn.data_type),
        });

        this.endScope();
    }

    // ── Main entry point ──────────────────────────────
    compile(ast) {
        this.emit(OpCode.OP_CALL);
        const mainAddrPos = this.emitPlaceholder();
        this.emit(0);
        this.emit(OpCode.OP_POP);
        this.emit(OpCode.OP_HALT);

        // Pass 1: register all functions
        for (let i = 0; i < ast.items.length; i++) {
            const fn = ast.items[i];
            if (fn.type !== NodeType.NODE_FUNC_DEF) {
                throw new CompileError(fn.line, 'only function definitions allowed at top level');
            }
            if (this.findFunc(fn.name) >= 0) {
                throw new CompileError(fn.line, `function '${fn.name}' already defined`);
            }
            this.registerFunc(fn.name, fn.items.length, fn.line);
        }

        // Pass 2: compile bodies
        for (let i = 0; i < ast.items.length; i++) {
            this.compileFunc(ast.items[i]);
        }

        // Resolve forward references
        for (let i = 0; i < this.patchCount; i++) {
            const fi = this.patches[i].funcIdx;
            const pos = this.patches[i].position;
            if (this.functions[fi].address < 0) {
                throw new CompileError(0, `function '${this.functions[fi].name}' declared but never defined`);
            }
            this.patchAt(pos, this.functions[fi].address);
        }

        // Patch main address
        const mainIdx = this.findFunc('main');
        if (mainIdx < 0) {
            throw new CompileError(0, "no 'main' function defined");
        }
        this.patchAt(mainAddrPos, this.functions[mainIdx].address);
        this.prog.entry = 0;

        return this.prog;
    }
}

/**
 * Disassemble bytecode into an array of instruction objects (for visualization).
 */
function disassemble(prog) {
    const instructions = [];
    let i = 0;
    while (i < prog.count) {
        const op = prog.code[i];
        const addr = i;
        const name = opName(op);
        let operands = '';
        let size = 1;

        switch (op) {
            case OpCode.OP_CONST:
            case OpCode.OP_LOAD:
            case OpCode.OP_STORE:
            case OpCode.OP_JMP:
            case OpCode.OP_JZ:
            case OpCode.OP_ENTER:
                operands = String(prog.code[i + 1]);
                size = 2;
                break;
            case OpCode.OP_CALL:
                operands = `addr=${prog.code[i + 1]} nargs=${prog.code[i + 2]}`;
                size = 3;
                break;
        }

        instructions.push({ addr, name, operands, op, size });
        i += size;
    }
    return instructions;
}

function compile(ast) {
    const c = new Compiler();
    const prog = c.compile(ast);
    return { program: prog, semanticInfo: c.semanticInfo };
}

if (typeof module !== 'undefined') {
    module.exports = { OpCode, opName, Program, Compiler, compile, disassemble };
}
